### PR TITLE
Fixed some spot bugs on azure cosmos SDK

### LINKED ISSUE #9044: [Cosmos v4] Evaluate byte array to string without charset/encoding
Missing encoding when converting bytes to string https://github.com/Azure/azure-sdk-for-java/blob/3e9b51beb5f5bc96f3e262445f256cea11108b8b/sdk/cosmos/azure-cosmos/src/main/java/com/azure/cosmos/implementation/RxGatewayStoreModel.java#L330 if service always returns error in UTF8 format, it has to be specified explicitly here.

### LINKED ISSUE #9054: [Cosmos v4] Spotbug: evaluate Double Checked Lock (DCL) in CosmosAsyncItemResponse::getItem
While scanning the code for spot-bug reported RCN_REDUNDANT_NULLCHECK_OF_NULL_VALUE, I've noticed that below method is using DCL which is not safe:

https://github.com/Azure/azure-sdk-for-java/blob/c880a2cbeac8cdf2ef8b58b797b35620d703475c/sdk/cosmos/azure-cosmos/src/main/java/com/azure/cosmos/models/CosmosAsyncItemResponse.java#L37 

https://www.cs.umd.edu/~pugh/java/memoryModel/DoubleCheckedLocking.html

### LINKED ISSUE #9078: [Cosmos v4] Evaluate switch without default branch
The following switch statements are without the default branch. Please evaluate that this is by design.

https://github.com/Azure/azure-sdk-for-java/blob/d888e8aa057d9675e357fde0dff2e68654b9fc0d/sdk/cosmos/azure-cosmos/src/main/java/com/azure/cosmos/implementation/routing/PartitionKeyInternal.java#L121

https://github.com/Azure/azure-sdk-for-java/blob/d888e8aa057d9675e357fde0dff2e68654b9fc0d/sdk/cosmos/azure-cosmos/src/main/java/com/azure/cosmos/implementation/routing/MurmurHash3_32.java#L49 

https://github.com/Azure/azure-sdk-for-java/blob/d888e8aa057d9675e357fde0dff2e68654b9fc0d/sdk/cosmos/azure-cosmos/src/main/java/com/azure/cosmos/implementation/directconnectivity/rntbd/RntbdRequestHeaders.java#L992

### LINKED ISSUE #9087: [Cosmos V4] evaluate the correctness of volatile variable increment
Volatile member variables are incremented. These increments don't guarantee atomicity for "read|increment|update". Double-check whether these increments are intended to be called from multiple threads if yes then switching to AtomicInteger is safe if no concurrency involved then volatile could be removed.

https://github.com/Azure/azure-sdk-for-java/blob/591d77d6260d9da8a148d1f1ffda2372c4a06143/sdk/cosmos/azure-cosmos/src/main/java/com/azure/cosmos/implementation/RetryPolicyWithDiagnostics.java#L54

https://github.com/Azure/azure-sdk-for-java/blob/591d77d6260d9da8a148d1f1ffda2372c4a06143/sdk/cosmos/azure-cosmos/src/main/java/com/azure/cosmos/implementation/changefeed/implementation/AutoCheckpointer.java#L64

https://github.com/Azure/azure-sdk-for-java/blob/591d77d6260d9da8a148d1f1ffda2372c4a06143/sdk/cosmos/azure-cosmos/src/main/java/com/azure/cosmos/implementation/query/DefaultDocumentQueryExecutionContext.java#L153

### LINKED ISSUE #9091: [Cosmos v4] Possible access to partially constructed object
The following access to itemsTypes member can result in accessing a partially constructed value.

```java
if (this.itemTypes == null) {
    synchronized (this) {
        if (this.itemTypes == null) {
            this.itemTypes = new ArrayList<ItemType>(result1.size());
            for (QueryItem item : result1) {
                this.itemTypes.add(ItemTypeHelper.getOrderByItemType(item.getItem()));
            }
        }
    }
}
```
Ref:
https://github.com/Azure/azure-sdk-for-java/blob/f770d24d9a41923298c86cbaf422241501086f1c/sdk/cosmos/azure-cosmos/src/main/java/com/azure/cosmos/implementation/query/orderbyquery/OrderbyRowComparer.java#L44

### LINKED ISSUE #9092: [Cosmos v4] Possible dropping of QueryMetrics
The add method in [QueryMetrics](https://github.com/Azure/azure-sdk-for-java/blob/master/sdk/cosmos/azure-cosmos/src/main/java/com/azure/cosmos/implementation/QueryMetrics.java) is  shown below:

It creates a new `QueryMetrics` instance which contains items in the current `QueryMetrics` (on which `add` is called) and given list of QueryMetrics (`queryMetricsArgs`) i.e. this method is not mutating the object on which `add()` is called:

```java
public QueryMetrics add(QueryMetrics... queryMetricsArgs) {
    ArrayList<QueryMetrics> queryMetricsList = new ArrayList<QueryMetrics>();
    for (QueryMetrics queryMetrics : queryMetricsArgs) {
        queryMetricsList.add(queryMetrics);
    }
    queryMetricsList.add(this);
    return QueryMetrics.createFromCollection(queryMetricsList);
}
```

OrderUtils.java::[apply](https://github.com/Azure/azure-sdk-for-java/blob/1300def7a6c810eb0b9f3a9fe175316e7f4f98f7/sdk/cosmos/azure-cosmos/src/main/java/com/azure/cosmos/implementation/query/OrderByUtils.java#L70)() method uses this method as:

```java
queryMetricsMap.get(key).add(qm);
```

Note that there is a `QueryMetrics` object in `queryMetricsMap` identified by `key`, on which it calls add() but this call has no effect given return value is ignored.

Similar use of `add` can be found here as well:

https://github.com/Azure/azure-sdk-for-java/blob/1300def7a6c810eb0b9f3a9fe175316e7f4f98f7/sdk/cosmos/azure-cosmos/src/main/java/com/azure/cosmos/implementation/query/AggregateDocumentQueryExecutionContext.java#L92

### LINKED ISSUE #9108: [Cosmos v4] Consider removing CancellationTokenSource::close
This [code](https://github.com/Azure/azure-sdk-for-java/blob/650ab28c669cf2132129abfd4da79007ae8f9e1a/sdk/cosmos/azure-cosmos/src/main/java/com/azure/cosmos/implementation/changefeed/CancellationTokenSource.java#L37)

```java
@Override
public synchronized void close() throws IOException {
    if (tokenSourceClosed) return;
}
```

Consider removing `CancellationTokenSource`'s inheritance from `Closeable`, if there is no cleanup to be done on `close()`.

### LINKED ISSUE #9179: [Cosmos v4] Evaluate redundant null check and un-called private method
Following suppressions are added based on [PR](https://github.com/Azure/azure-sdk-for-java/pull/9111/) discussion
```
  <Match>
    <Class name="com.azure.cosmos.implementation.SessionContainer"/>
    <Method name="setSessionToken"/>
    <Bug pattern="RCN_REDUNDANT_NULLCHECK_WOULD_HAVE_BEEN_A_NPE"/>
  </Match>

  <Match>
    <Class name="com.azure.cosmos.implementation.SessionContainer"/>
    <Method name="shouldUpdateSessionToken"/>
    <Bug pattern="RCN_REDUNDANT_NULLCHECK_WOULD_HAVE_BEEN_A_NPE"/>
  </Match>

  <Match>
    <Class name="com.azure.cosmos.implementation.routing.MurmurHash3_128"/>
    <Method name="rotateLeft32"/>
    <Bug pattern="UPM_UNCALLED_PRIVATE_METHOD"/>
  </Match>
```

### PR DESCRIPTION
**Pull request by kushagraThapar**
* Fixes #9091
* Fixes #9092
* Fixes #9108
* Fixes #9179
* Fixes #9054
* Fixes #9078
* Fixes #9044
* Fixes #9087

### CODE REVIEW

File: sdk/cosmos/azure-cosmos/src/main/java/com/azure/cosmos/implementation/query/AggregateDocumentQueryExecutionContext.java [hunk 1]
```diff
@@ -84,7 +84,7 @@ public Flux<FeedResponse<Document>> drainAsync(int maxPageSize) {
                         for(String key : BridgeInternal.queryMetricsFromFeedResponse(page).keySet()) {
                             if (queryMetricsMap.containsKey(key)) {
                                 QueryMetrics qm = BridgeInternal.queryMetricsFromFeedResponse(page).get(key);
-                                queryMetricsMap.get(key).add(qm);
+                                queryMetricsMap.put(key, QueryMetrics.addQueryMetrics(queryMetricsMap.get(key), qm));
```
Comment (xinlian12) [inline sdk/cosmos/azure-cosmos/src/main/java/com/azure/cosmos/implementation/query/AggregateDocumentQueryExecutionContext.java:87]: should we use compute()

**Review by xinlian12 — APPROVED** [review]
LGTM, thanks

File: sdk/cosmos/azure-cosmos/src/main/java/com/azure/cosmos/implementation/query/AggregateDocumentQueryExecutionContext.java (hunk 1 above)
Comment (kushagraThapar) [reply sdk/cosmos/azure-cosmos/src/main/java/com/azure/cosmos/implementation/query/AggregateDocumentQueryExecutionContext.java]: Good point, done, and some more optimizations.

**Review by xinlian12 — APPROVED** [review]
LGTM, thanks

**Pull request merged**